devito/symbolics.py

```diff
@@ -82,90 +92,206 @@ def dse_indexify(expr):
-    for i in range(len(expr)):
-        stencils[i] = Eq(expr[i].lhs, stencils[i].rhs)
 
-    to_revert = {}
-    to_keep = []
+class Rewriter(object):
 
-    # Restores IndexedBases if they are collected by CSE and
-    # reverts changes to simple index operations (eg: t - 1)
-    for temp, value in temps:
-        if isinstance(value, IndexedBase):
-            to_revert[temp] = value
-        elif isinstance(value, Indexed):
-            to_revert[temp] = value
-        elif isinstance(value, Add) and not \
-                set([t, x, y, z]).isdisjoint(set(value.args)):
-            to_revert[temp] = value
-        else:
-            to_keep.append((temp, value))
-
-    # Restores the IndexedBases and the Indexes in the assignments to revert
-    for temp, value in to_revert.items():
-        s_dict = {}
-        for arg in preorder_traversal(value):
-            if isinstance(arg, Indexed):
-                new_indices = []
-                for index in arg.indices:
-                    if index in to_revert:
-                        new_indices.append(to_revert[index])
-                    else:
-                        new_indices.append(index)
-                if arg.base.label in to_revert:
-                    s_dict[arg] = Indexed(to_revert[value.base.label], *new_indices)
-        to_revert[temp] = value.xreplace(s_dict)
-
-    subs_dict = {}
-
-    # Builds a dictionary of the replacements
-    for expr in stencils + [assign for temp, assign in to_keep]:
-        for arg in preorder_traversal(expr):
-            if isinstance(arg, Indexed):
-                new_indices = []
-                for index in arg.indices:
-                    if index in to_revert:
-                        new_indices.append(to_revert[index])
-                    else:
-                        new_indices.append(index)
-                if arg.base.label in to_revert:
-                    subs_dict[arg] = Indexed(to_revert[arg.base.label], *new_indices)
-                elif tuple(new_indices) != arg.indices:
-                    subs_dict[arg] = Indexed(arg.base, *new_indices)
-            if arg in to_revert:
-                subs_dict[arg] = to_revert[arg]
-
-    stencils = [stencil.xreplace(subs_dict) for stencil in stencils]
-
-    to_keep = [Eq(temp[0], temp[1].xreplace(subs_dict)) for temp in to_keep]
-
-    # If the RHS of a temporary variable is the LHS of a stencil,
-    # update the value of the temporary variable after the stencil
-
-    new_stencils = []
-
-    for stencil in stencils:
-        new_stencils.append(stencil)
-
-        for temp in to_keep:
-            if stencil.lhs in preorder_traversal(temp.rhs):
-                new_stencils.append(temp)
-                break
-
-    return to_keep + new_stencils
+    """
+    Transform expressions to reduce their operation count.
+    """
+
+    # Do more factorization sweeps if the expression operation count is
+    # greater than this threshold
+    FACTORIZER_THS = 15
```

Comment:
The factorisation engine is makes use of heuristics, so recursively doing more sweeps may result in a better operation count reduction. However, it might be the case that for insane stencils (dunno, say TTI order 14-16?), the symbolic processing time becomes too high if we apply factorisation recursively to all expressions performing more than ``FACTORIZER_THS`` flops (not only do the temporaries become individually more expensive, but also the total number of temporaries > ``FACTORIZER_THS`` increases!). So this is a knob that we might want to expose (somehow), in the future, to work around this problem. A long symbolic processing time wouldn't be a problem when we do long runs and we care about performance (these are supposed to last for many hours or days), although *it might be* a problem at development or debugging time.

I don't think there's an urgency to expose this parameter in this PR